src/Microsoft.ML.TensorFlow/TensorFlow/TensorflowUtils.cs

```diff
@@ -20,51 +20,53 @@ public static class TensorFlowUtils
                 var mlType = Tf2MlNetTypeOrNull(tfType);
 
                 // If the type is not supported in ML.NET then we cannot represent it as a column in an ISchema.
                 // We also cannot output it with a TensorFlowTransform, so we skip it.
                 if (mlType == null)
                     continue;
 
-                var shape = graph.GetTensorShape(op[0]);
-                var shapeArray = shape.ToIntArray();
-
-                inputOpsLengths.Add(op.NumInputs);
-                MetadataUtils.MetadataGetter<VBuffer<ReadOnlyMemory<char>>> inputOpsGetter = null;
+                // Construct the final ML.NET type of a Tensorflow variable.
+                var shapeArray = graph.GetTensorShape(op[0]).ToIntArray();
+                var columnType = new VectorType(mlType);
+                if (Utils.Size(shapeArray) > 0 && shapeArray.Skip(1).All(x => x > 0))
+                    columnType = new VectorType(mlType, shapeArray[0] > 0 ? shapeArray : shapeArray.Skip(1).ToArray());
+
+                // There can be at most two metadata fields.
+                //  1. The first field always presents. Its value is this operator's ID. For example,
+                //     if an output is produced by an operator named "A", the value of this field should be "A".
+                //  2. The second field stores operators whose outputs are consumed by this operator. In other words,
+                //     these values are names of some upstream operators which should be evaluated before executing
+                //     the current operator. It's possible that one operator doesn't need any input, so this field
+                //     can be missing.
+                var metadataBuilder = new MetadataBuilder();
+                metadataBuilder.Add(OpType, TextType.Instance, (ref ReadOnlyMemory<char> value) => value = op.OpType.AsMemory());
                 if (op.NumInputs > 0)
-                {
-                    var inputOps = new ReadOnlyMemory<char>[op.NumInputs];
-                    for (int i = 0; i < op.NumInputs; i++)
-                    {
-                        var input = op.GetInput(i);
-                        inputOps[i] = new ReadOnlyMemory<char>(input.Operation.Name.ToArray());
-                    }
-                    inputOpsGetter = (int col, ref VBuffer<ReadOnlyMemory<char>> dst) =>
-                        dst = new VBuffer<ReadOnlyMemory<char>>(op.NumInputs, inputOps);
-                }
-                inputOpsGetters.Add(inputOpsGetter);
-
-                MetadataUtils.MetadataGetter<ReadOnlyMemory<char>> opTypeGetter =
-                    (int col, ref ReadOnlyMemory<char> dst) => dst = new ReadOnlyMemory<char>(op.OpType.ToArray());
-                opTypeGetters.Add(opTypeGetter);
-
-                var columnType = Utils.Size(shapeArray) == 1 && shapeArray[0] <= 0 ? new VectorType(mlType) :
```

Comment:
Note: we should not use recursive if-else-like at all. It's not readable. In fact, this line is equivalent to
```charp
                var columnType = new VectorType(mlType);
                if (Utils.Size(shapeArray) > 0 && shapeArray.Skip(1).All(x => x > 0))
                    columnType = new VectorType(mlType, shapeArray[0] > 0 ? shapeArray : shapeArray.Skip(1).ToArray());
```